core/threaded/scheduler_GEDF_NP.c

```diff
@@ -0,0 +1,368 @@
+/* Global Earliest Deadline First (GEDF) non-preemptive scheduler for the
+threaded runtime of the C target of Lingua Franca. */
+
+/*************
+Copyright (c) 2022, The University of Texas at Dallas.
+Copyright (c) 2022, The University of California at Berkeley.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************/
+
+/** 
+ * Global Earliest Deadline First (GEDF) non-preemptive scheduler for the
+ * threaded runtime of the C target of Lingua Franca.
+ *  
+ * @author{Soroush Bateni <soroush@utdallas.edu>}
+ * @author{Edward A. Lee <eal@berkeley.edu>}
+ * @author{Marten Lohstroh <marten@berkeley.edu>}
+ */
+
+#ifndef NUMBER_OF_WORKERS
+#define NUMBER_OF_WORKERS 1
+#endif // NUMBER_OF_WORKERS
+
+#include "scheduler.h"
+#include "../platform.h"
+#include "../utils/semaphore.c"
+#include "../utils/pqueue_support.h"
+#include "scheduler_sync_tag_advance.c"
+#include <assert.h>
+
+#ifndef MAX_REACTION_LEVEL
+#define MAX_REACTION_LEVEL INITIAL_REACT_QUEUE_SIZE
+#endif
+
+/////////////////// External Variables /////////////////////////
+extern lf_mutex_t mutex;
+
+/////////////////// Scheduler Variables and Structs /////////////////////////
+/**
+ * @brief Atomically keep track of how many worker threads are idle.
```

Comment:
> One thread going idle can change _lf_sched_semaphore->count by more than one.

Yes, the comment on the semaphore is incorrect. The semaphore doesn't actually keep track of the number of idle worker threads. The semaphore will be released by the scheduler with the count being the _maximum_ number of worker threads that can wake up (with the optimization that the calling worker thread doesn't need a notification, so it would be excluded from the semaphore count). If one worker thread is fast enough to go through all the reactions before other worker threads can wake up, that is not an error condition. This is a best-effort method of work distribution. In this case, a greedy worker is not necessarily a problem. This greediness is inherent in the fact that all worker threads share the same work queue. A fast worker thread can go through all the work before others get the chance.

On the other hand, the PEDF scheduler has a more direct method of waking up specific worker threads because each worker thread has its own work queue (although it does have work-stealing if it runs out of work).


> It looks like the semaphore is not providing more functionality than the lf_cond_t that it uses, because it ensures that workers block when they finish and the reaction queue is empty, and it ensures that they unblock when _lf_sched_notify_workers is invoked.

This is not true. The semaphore is in charge of controlling the _maximum_ (not exact) number of worker threads that can wake up. This would not be trivially possible using a `lf_cond_t`.

Reply:
Thanks for your explanation! I'll have to re-read the code now that I know what the semaphore signifies.

Reply:
I updated the comment to better reflect this point.